gossip/src/cluster_info.rs

```diff
@@ -1783,89 +1783,86 @@ impl ClusterInfo {
+                {
+                    // Log contact info
+                    info!(
+                        "\n{}\n\n{}",
+                        self.contact_info_trace(),
+                        self.rpc_info_trace()
+                    );
+                    last_contact_info_trace = start;
                 }
-                let mut generate_pull_requests = true;
-                loop {
-                    let start = timestamp();
-                    if self.contact_debug_interval != 0
-                        && start - last_contact_info_trace > self.contact_debug_interval
-                    {
-                        // Log contact info
-                        info!(
-                            "\n{}\n\n{}",
-                            self.contact_info_trace(),
-                            self.rpc_info_trace()
-                        );
-                        last_contact_info_trace = start;
-                    }
 
-                    if self.contact_save_interval != 0
-                        && start - last_contact_info_save > self.contact_save_interval
-                    {
-                        self.save_contact_info();
-                        last_contact_info_save = start;
-                    }
+                if self.contact_save_interval != 0
+                    && start - last_contact_info_save > self.contact_save_interval
+                {
+                    self.save_contact_info();
+                    last_contact_info_save = start;
+                }
 
-                    let (stakes, _feature_set) = match bank_forks {
-                        Some(ref bank_forks) => {
-                            let root_bank = bank_forks.read().unwrap().root_bank();
-                            (
-                                root_bank.staked_nodes(),
-                                Some(root_bank.feature_set.clone()),
-                            )
-                        }
-                        None => (Arc::default(), None),
-                    };
-                    let _ = self.run_gossip(
-                        &thread_pool,
-                        gossip_validators.as_ref(),
+                let (stakes, _feature_set) = match bank_forks {
+                    Some(ref bank_forks) => {
+                        let root_bank = bank_forks.read().unwrap().root_bank();
+                        (
+                            root_bank.staked_nodes(),
+                            Some(root_bank.feature_set.clone()),
+                        )
+                    }
+                    None => (Arc::default(), None),
+                };
+                let _ = self.run_gossip(
+                    &thread_pool,
+                    gossip_validators.as_ref(),
+                    &recycler,
+                    &stakes,
+                    &sender,
+                    // Generate pull requests once every 10 gossip rounds.
+                    (round % 10) == 0,
```

Comment:
we already decreased the number of pull requests in [PR: #33719](https://github.com/solana-labs/solana/pull/33719). And now we're decreasing them again by 5x. I understand that the updated push message logic is much better than it was before. But I have a couple questions:
1) How did you land on once every 10 gossip rounds? 
2) With a further reduction in Pull Requests are we sure CrdsValues are propagating via Push well enough that nodes are getting the data they need in time?

Reply:
The ultimate goal is that pull requests should only serve as a mechanism for:

1. initial crds table sync when the node boots up.
2. mitigation against transient network partitions.
3. defense against eclipse attacks.

Otherwise, we really shouldn't rely on pull requests/responses for message propagation. The computational overhead on both the sender and the receiver is pretty high and it is partly proportional to CRDS table size. The bandwidth requirement for requests is also a function of CRDS table size. So even if messages are already propagated decently from the push path, and there are only a few messages that you miss, you still have to spend a computational cost and network bandwidth which becomes huge as the gossip table becomes bigger.

Also propagating messages through pull request/response is pretty slow as it requires a round-trip packet and hefty computations on both side.

If we need to improve message propagation it is just cheaper to add more redundancy to push mechanism rather than sending pull requests. For example we can increase push fanout or `min_ingress_{nodes,stake}` when pruning:
https://github.com/solana-labs/solana/blob/c6c23405d/gossip/src/received_cache.rs#L93-L124 
On a relative basis, that has less overhead than sending and processing pull request/responses.

Obviously we need to debug and improve push mechanism with the same bandwidth requirements first before adding more redundancy to it as a last resort.

> 1. How did you land on once every 10 gossip rounds?

10 times `GOSSIP_SLEEP_MILLIS` (100ms) is 1 second, so it will be once every second which I thought shouldn't be too bad.

> 2. With a further reduction in Pull Requests are we sure CrdsValues are propagating via Push well enough that nodes are getting the data they need in time?

* That is something we need to monitor and evaluate on testnet and also once this change reaches mainnet.
* But the thing is if propagation is not good, as explained above, we shouldn't utilize or rely on pull requests to fix that, but instead debug and fix the push path. Using pull requests to propagate messages is pretty awful.
* I am also expecting making pull requests less frequent would also help to better inspect and understand push path propagation, because currently pull requests are effectively a backdoor in the propagation graph which disguises where things are not working and need to be improved.

Reply:
lots of good points here. thanks for taking the time to explain. Ya totally understand we don't want to rely on pull messages really at all. 

My concern was just that we rely on vote propagation through gossip for optimistic confirmation. so just wanted to make sure that removing Pull Requests isn't going to affect this since votes may not be seen as widely. but I guess it already takes a lot of time to get messages via Pull Request so not sure how much votes received via Pull Request help in OC. 

> If we need to improve message propagation it is just cheaper to add more redundancy to push mechanism rather than sending pull requests. 

Ya totally agree. good point. 

> 10 times GOSSIP_SLEEP_MILLIS (100ms) is 1 second, so it will be once every second which I thought shouldn't be too bad.

Makes sense. i guess we'll just keep an eye on it as it gets rolled out

> But the thing is if propagation is not good, as explained above, we shouldn't utilize or rely on pull requests to fix that, but instead debug and fix the push path. Using pull requests to propagate messages is pretty awful.

True. just wanted to make sure we don't scale back too quickly without fully knowing 1) how well push messages distribute messages and 2) how well OC and other components will perform if we scale back Pull Requests. But 100% agree with the sentiment that we need to get rid of Pull Requests as much as possible. 

> I am also expecting making pull requests less frequent would also help to better inspect and understand push path propagation, because currently pull requests are effectively a backdoor in the propagation graph which disguises where things are not working and need to be improved.

Ya 100p agree. good point